### ISSUE TITLE
Question: Thread-Safety

### ISSUE THREAD

**Issue by christophwille** [bytecodealliance/wasmtime-dotnet#90]
I have asked that in the past most likely via an at-mention in https://github.com/christophwille/dotnet-opa-wasm/ - but today it came back up for multi-threaded usage of the Wasmtime API. I have cooked up a sample and I am pretty sure I remember our past discussion wrongly... the overall goal is to not incur the compilation hit.

https://github.com/christophwille/dotnet-opa-wasm/blob/master/spikes/HigherLevelApisSpike/IPolicyFactory.cs#L129

I load a wasm file into a byte array, create an engine and a use the engine for module loadt - and for the sake of discussion, throw away that engine. The question now is: is Module safe to use in parallel later for 

// _store is created on a new Engine, _linker same
var instance = _linker.Instantiate(_store, module);

Ie can one cached Module be used to create unlimited instances *in parallel*.

Thanks.

**Comment by peterhuene** [comment]
Yes, you should be able to share a _module_ between threads for the purpose of instantiation. The module will keep a reference on the engine used to create it, but you can dispose of your own reference to that engine if it is no longer needed.

It's only `Store` (and all store-related objects like instances) that cannot be shared between threads.

Also, the .NET API supports loading of modules from "precompiled" modules, but not creating them as we don't currently expose the `Engine::precompile_module` method fro the Wasmtime API for the .NET API to use. If you have Wasmtime installed, you can use `wasmtime compile <module>` to generate a precompiled file (`cwasm`), however.

The benefit of using a precompiled file is that there is no JIT compilation at all; the module is simply mapped into memory upon load.

**Comment by christophwille** [comment]
Ah, didn't know about the precompilation feature, interesting though (very much so).

Then my current object model was already correctly split between Module and instance with linker and store. And for my new approach I only need to cache one Module for creating multiple instances.

**Comment by peterhuene** [comment]
From a cursory inspection, that should work as you intend.

**Comment by willhausman** [comment]
> Yes, you should be able to share a _module_ between threads for the purpose of instantiation. The module will keep a reference on the engine used to create it, but you can dispose of your own reference to that engine if it is no longer needed.

@peterhuene does this mean that it doesn't matter if the same `Engine` is used to create an `Instance` from a `Module`?

e.g.

```csharp
var engine = new Engine();
var module = Module.FromFile(engine, file);

// now stuff happens with the application keeping module in memory but disposing engine

var engine = new Engine();
var linker = new Linker(engine);
var store = new Store(engine);
DoSetupLinkerStuff();
var instance = linker.Instantiate(store, module);  // <- different engines
```

**Comment by peterhuene** [comment]
Cross-engine instantiation isn't supported and I'd expect that to throw an exception (at least it should; please file a bug if not).

You'll need to use the same engine for the module, linker, and store. I forgot that we don't have an `Engine` property on `Module` or `Linker` in the .NET API like we do in the underlying Wasmtime API in Rust as it's not exposed via the C bindings.

You'll need to keep a reference to the same engine around for now.

**Comment by christophwille** [comment]
> You'll need to use the same engine for the module, linker, and store. I forgot that we don't have an `Engine` property on `Module` or `Linker` in the .NET API like we do in the underlying Wasmtime API in Rust as it's not exposed via the C bindings.

Could you expose such a property? Or is this a no-no for the .NET object model? Because with that would a allow

`csharp
using var opaPolicy = new OpaPolicy(module.Runtime, module);`

without keeping the Engine around "some other way". The Engine is thread-safe though?

**Comment by christophwille** [comment]
> Cross-engine instantiation isn't supported and I'd expect that to throw an exception (at least it should; please file a bug if not).

https://github.com/christophwille/dotnet-opa-wasm/blob/master/src/Opa.Wasm.UnitTests/ThreadSafetyTests.cs#L10 it does fail but not very "obvious" as to the cause unless you look at the line it is failing on https://github.com/christophwille/dotnet-opa-wasm/blob/master/src/Opa.Wasm/OpaPolicy.cs#L140 and know the innards of Wasmtime.

**Comment by christophwille** [comment]
Kind of related question: if Engine and Module are tied tightly together - is there somewhere guidance on tradeoffs/use cases for (a) one engine married to one module (possibly hundreds around at the same time) vs (b) one engine for multiple modules? Which approach is "better", to use when?

**Comment by peterhuene** [comment]
Right now there isn't a way for the .NET API to expose an `Engine` property on either `Module` or `Linker` without storing a reference on the .NET side which is generally something I've been trying to avoid doing (I prefer the native side keep the references to keep things simpler and relatively deterministic).

It wouldn't be hard to add C API functions to get a reference to the engine for a given module or linker and then make use of those from the .NET API, though. I can look at doing that when I get a chance.

With regards to the exception you're getting, that's definitely an unexpected and confusing one, so I can dig into why it isn't the expected "inter-engine instantiation isn't supported" message.

**Comment by peterhuene** [comment]
Oh and `Engine` is thread safe too.

**Comment by peterhuene** [comment]
It's generally fine to create a single `Engine` per process and create all modules from that engine.

An engine is simply a JIT compiler, an instance allocator, and a signature registry.

The JIT compiler is thread-safe and will synchronize the compilation of any modules being compiled for that engine.

The instance allocator is what is used to allocate new instances from the instantiation of modules. By default (and what's supported by the .NET API), it's an on-demand allocator that simply allocates the needed resources of an instance when requested; it doesn't store any state that grows over time. The Rust API has another allocator type that uses resource pools to allocate from, but that's not yet supported by .NET.

Lastly, the signature registry is a global place we assign unique ids to function signatures. This is used as an optimization so that we can very quickly verify function signatures when doing indirect calls from WebAssembly. This will grow over time, but is bounded by how many unique function signatures are encountered in compiled modules and host functions (thus likely a fairly small set).

None of these things would be impacted from having a single `Engine` shared across all modules.

**Comment by alc-aardvark** [comment]
I have tried to run a WASM module from a C# unit test where there are 2 threads calling the module in a loop.  This dies with one of the threads ending in with the following exception

**`wasm trap: call stack exhausted`**

I count the iterations and the thread with the above exception calls the WASM module method 89,498 times, while the other thread calls the WASM module method 0 times.

It is like the WASM module method is being called repeatedly by the one thread but never cleans up the call stack.

If I run this as 1 thread OR put a C# Lock {} block around the WASM method call, then it all runs fine, and I have called the method for over 5,000,000 iterations without an issue.

The WASM module was written in C and and generated using Emscripten.  The method takes in two string (pointers) and a an Int64 and returns a string (pointer).

The memory is being Free'ed.

**C method** 
`char* TryVerify(char* token, char* originHeader, uint64_t serverUtcUnixNowMilliseconds)`

**C# call**
```
        object lockObject = new object();
        public bool TryVerify(string token, string originHeader, out TokenData tokenData)
        {
            //lock (lockObject)
            //{
                var serverUtcUnixNowMilliseconds = new DateTimeOffset(DateTime.UtcNow).ToUnixTimeMilliseconds();
                int inPtrToken = WriteStringToUTF8(token);
                int inPtrOriginHeader = WriteStringToUTF8(originHeader);
                int outPtr = _TryVerify.Invoke(inPtrToken, inPtrOriginHeader, serverUtcUnixNowMilliseconds);
                var json = Memory.ReadNullTerminatedString(outPtr);
                _Free(outPtr);

                tokenData = System.Text.Json.JsonSerializer.Deserialize<TokenData>(json, new JsonSerializerOptions() { PropertyNameCaseInsensitive = true }) ?? throw new Exception("Failed to deserialize token data from wasm");
                if (tokenData == null)
                {
                    tokenData = new TokenData();
                }

                return tokenData.IsOk;
           // }
        }

```

**Comment by alc-aardvark** [comment]
@peterhuene should I raise my issue above as a new issue?
